### PR TITLE
Add BlockDestroyedByNeighborEvent

### PR DESCRIPTION
**Pull request by Cryptite**
This PR adds the event by the same name but also serves a second purpose. There is a new variable called pendingPlayerBlockEvents which serves to help propagate which Player cause a block break which cause any subsequent block breaks.

Examples of BlockDestroyedByNeighborEvent are things like breaking a column of vines, bamboo, DoubleTall plant blocks or any block that is broken as a result of an initial block break.

The idea is that this can be expanded up by other contributors to add Player to more Events that currently have none. An example of this in a following PR would be SpongeAbsorbEvent, which will be able to have which Player placed said Sponge block, etc.

### CODE REVIEW

File: Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+ // CraftBukkit end
+ 
+@@ -102,6 +105,28 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+     public final Map<Explosion.CacheKey, Float> explosionDensityCache = new HashMap<>(); // Paper - Optimize explosions
+     public java.util.ArrayDeque<BlockRedstoneTorch.RedstoneUpdateInfo> redstoneUpdateInfos; // Paper - Move from Map in BlockRedstoneTorch to here
+ 
++    // Paper start - Abstract holder class that can be used to track what Player did the last "block thing"
++    public static class PendingBlockEvent {
++
++        public final BlockPosition block;
```
Comment (Proximyst) [inline Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch]: Can't this be modified and just requeue the event?

File: Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+             // they are NOT used with same intent and the above should not fire this event. The above method is more of a BlockSetToAirEvent,
+@@ -645,8 +680,23 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+     public void a(BlockPosition blockposition, Block block, BlockPosition blockposition1) {
+         if (!this.isClientSide) {
+             IBlockData iblockdata = this.getType(blockposition);
++            org.bukkit.block.Block blockAt = world.getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()); // Paper
+ 
+             try {
++                // Paper start
++                if (blockAt.getType() != org.bukkit.Material.AIR && iblockdata.getMaterial() == Material.AIR) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch]: `Material#isAir` is going to be *vital* to be ready for the 1.17 update.

File: Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+--- a/src/main/java/net/minecraft/server/BlockBase.java
++++ b/src/main/java/net/minecraft/server/BlockBase.java
+@@ -525,6 +525,19 @@ public abstract class BlockBase {
+ 
+                 blockposition_mutableblockposition.a((BaseBlockPosition) blockposition, enumdirection);
+                 IBlockData iblockdata = generatoraccess.getType(blockposition_mutableblockposition);
++
++                // Paper start - Propagate the pendingPlayerBlockEvents
++                World world = generatoraccess.getMinecraftWorld();
++                if (getMaterial() == Material.AIR && !iblockdata.canPlace(generatoraccess, blockposition_mutableblockposition)) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch]: `Material#isAir` is going to be *vital* to be ready for the 1.17 update.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch (hunk 3 above)
Comment (Cryptite) [reply Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch]: nms Material doesn't have an isAir check. Not sure what should be done here.

**Comment by Cryptite** [comment]
@Proximyst You told me to do it!

File: Spigot-API-Patches/0219-Add-BlockDestroyedByNeighborEvent.patch [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
++ * Called when a block is broken another block. This is generally the result of BlockPhysicsEvent propagation
++ */
++public class BlockDestroyedByNeighborEvent extends BlockEvent implements Cancellable {
++    private static final HandlerList handlers = new HandlerList();
++    private final Player player;
++    private final Block sourceBlock;
++    private boolean cancel;
++
++    public BlockDestroyedByNeighborEvent(@NotNull final Block theBlock, @Nullable Player player, @NotNull final Block sourceBlock) {
++        super(theBlock);
```
Comment (Octol1ttle) [inline Spigot-API-Patches/0219-Add-BlockDestroyedByNeighborEvent.patch:32]: Rename `theBlock` to `block`

File: Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -28,8 +28,11 @@ import org.bukkit.Bukkit;
+ import org.bukkit.craftbukkit.CraftServer;
+ import org.bukkit.craftbukkit.CraftWorld;
+ import org.bukkit.craftbukkit.block.CapturedBlockState;
++import org.bukkit.craftbukkit.block.CraftBlock;
+ import org.bukkit.craftbukkit.block.CraftBlockState;
+ import org.bukkit.craftbukkit.block.data.CraftBlockData;
++import org.bukkit.entity.Player;
++import org.bukkit.event.block.BlockDestroyedByNeighborEvent;
```
Comment (Octol1ttle) [inline Spigot-Server-Patches/0551-Add-BlockDestroyedByNeighborEvent.patch:112]: `// Paper` comments

File: Spigot-API-Patches/0219-Add-BlockDestroyedByNeighborEvent.patch [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
++
++import org.bukkit.block.Block;
++import org.bukkit.entity.Player;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Called when a block is broken another block. This is generally the result of BlockPhysicsEvent propagation
```
Comment (Octol1ttle) [inline Spigot-API-Patches/0219-Add-BlockDestroyedByNeighborEvent.patch:23]:
```suggestion
+ * Called when a block has broken another block. This is generally the result of BlockPhysicsEvent propagation
```

File: Spigot-API-Patches/0219-Add-BlockDestroyedByNeighborEvent.patch [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
++     */
++    @Nullable
++    public Player getPlayer() {
++        return player;
++    }
++
++    /**
++     * Gets the source block that caused this block break
++     *
++     * @return The Source Block which block is involved in this event
```
Comment (Octol1ttle) [inline Spigot-API-Patches/0219-Add-BlockDestroyedByNeighborEvent.patch:51]:
```suggestion
+     * @return the Source Block which broke the block
```

**Review by Octol1ttle — CHANGES_REQUESTED** [review]
Also a reminder to fix conflicts

**Pull request closed without merging**